draft-ietf-quic-qpack.md

```diff
@@ -270,184 +493,34 @@ in a single pass and include references to entries added while processing this
-
-For the encoder to identify which dynamic table entries can be safely used
-without a stream becoming blocked, the encoder tracks the absolute index of the
-decoder's Largest Known Received entry.
-
-When blocking references are permitted, the encoder uses acknowledgement of
-header blocks to identify the Largest Known Received index, as described in
-{{header-acknowledgement}}.
-
-To acknowledge dynamic table entries which are not referenced by header blocks,
-for example because the encoder or the decoder have chosen not to risk blocked
-streams, the decoder sends a Table State Synchronize instruction (see
-{{table-state-synchronize}}).
-
-# Conventions and Definitions
-
-The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
-document are to be interpreted as described in BCP 14 {{!RFC2119}} {{!RFC8174}}
-when, and only when, they appear in all capitals, as shown here.
-
-Definitions of terms that are used in this document:
-
-Header field:
-
-: A name-value pair sent as part of an HTTP message.
-
-Header list:
-
-: The ordered collection of header fields associated with an HTTP message.  A
-  header list can contain multiple header fields with the same name.  It can
-  also contain duplicate header fields.
-
-Header block:
-
-: The compressed representation of a header list.
-
-Encoder:
-
-: An implementation which transforms a header list into a header block.
-
-Decoder:
-
-: An implementation which transforms a header block into a header list.
-
-QPACK is a name, not an acronym.
-
-## Notational Conventions
-
-Diagrams use the format described in Section 3.1 of {{?RFC2360}}, with the
-following additional conventions:
-
-x (A)
-: Indicates that x is A bits long
-
-x (A+)
-: Indicates that x uses the prefixed integer encoding defined in Section 5.1 of
-  [RFC7541], beginning with an A-bit prefix.
-
-x ...
-: Indicates that x is variable-length and extends to the end of the region.
-
-#  Configuration
-
-QPACK defines two settings which are included in the HTTP/QUIC SETTINGS frame.
-
-  SETTINGS_HEADER_TABLE_SIZE (0x1):
-  : An integer with a maximum value of 2^30 - 1.  The default value is 4,096
-    bytes.  See {{table-dynamic}} for usage.
-
-  SETTINGS_QPACK_BLOCKED_STREAMS (0x7):
-  : An integer with a maximum value of 2^16 - 1.  The default value is 100.  See
-    {{overview-hol-avoidance}}.
-
-# Wire Format
-
-QPACK instructions occur in three locations, each of which uses a separate
-instruction space:
-
- - The encoder stream is a unidirectional stream of type `0x48` (ASCII 'H')
-   which carries table updates from encoder to decoder. Instructions on this
-   stream modify the dynamic table state without generating output to any
-   particular request.
+~~~~~ drawing
+             Base Index
+                 |
+                 V
+    +---+-----+-----+-----+-----+
+    | n | n-1 | n-2 | ... | d+1 |  Absolute Index
```

Comment:
This could be clearer, I think.

The 2 here (in n-2) is confusing.  What you really want is a diagram that shows the different values in a block.

Largest reference will be the largest value, that's your baseline.
Base offset is taken relative to the largest reference.  Here it's 2, which is confusing.  We need to show examples of where base allows for post-base references and one where it doesn't.

Including entries that have been evicted in this diagram confuses the matter.  It would be better to rely on text or to show that new entries appear on the left and old entries are evicted on the right.

Does this help?

```
Indexing Overview

    <- Absolute Indices (increase to left)
          L = Largest Reference
          B = Base Reference

       Newer                     Older
    <- Entries                 Entries ->
       Added                   Evicted

Indexing (simple case, B == L):
       ---+---+---+---+---+---+---+---+---
       ...| L |L-1|L-2|L-3|L-4|L-5|L-6|...
       ---+---+---+---+---+---+---+---+---
Index:      0   1   2   3   4   5   6  ...

Indexing (B > L):
       ---+---+---+---+---+---+---+---+---
       ...| B |B-1|...| L |L-1|L-2|L-3|...
       ---+---+---+---+---+---+---+---+---
Index:      x   x   x   D  D+1 D+2 D+3 ...
            x = invalid index
                        D = B - L

Post-Base Indexing (where B < L)
       ---+---+---+---+---+---+---+---+---
       ...| L |...|B+2|B+1| B |B-1|B-2|...
       ---+---+---+---+---+---+---+---+---
Index:                      0   1   2  ...
Post-Base:  P  ...  1   0
            P = L - B - 1
```